File: foo.ml
type nat =
| O
| S of nat

(** val plus : nat -> nat -> nat **)

let rec plus n m =
  match n with
  | O -> m
  | S p -> S (plus p m)

(** val mult : nat -> nat -> nat **)

let rec mult n m =
  match n with
  | O -> O
  | S p -> plus m (mult p m)

(** val foo : nat -> nat -> nat **)

let foo x y =
  plus (mult x x) y

File: foo.scala
object CoqMain {
sealed abstract class Nat
case class O() extends Nat
case class S(x1: Nat) extends Nat
def plus : (Nat => (Nat => Nat)) = {
(n:Nat) => (m:Nat) =>
n match {
case O() => m
case S(p) => S(plus(p)(m))
}
}
def mult : (Nat => (Nat => Nat)) = {
(n:Nat) => (m:Nat) =>
n match {
case O() => O()
case S(p) => plus(m)(mult(p)(m))
}
}
def foo : (Nat => (Nat => Nat)) = {
(x:Nat) => (y:Nat) =>
plus(mult(x)(x))(y)
}
}
